### ISSUE TITLE
Allow local .js file to be typed by its corresponding .d.ts file

### ISSUE THREAD

**Issue by balupton** [Microsoft/TypeScript#29056]
## Suggestion

If I have say `source/index.js` which has:

``` javascript
function isObject(value) {
	// null is object, hence the extra check
	return value !== null && typeof value === 'object'
}
```

And then I supply a `source/index.d.ts` file which has:

``` javascript
/** Checks to see if a value is an object */
export function isObject(value: any): boolean
```

Then I should not receive the following error in `source/index.js`:

```
[ts] Parameter 'value' implicitly has an 'any' type. [7006]
(parameter) value: any
```

And I should get complete intellisense.

## Use Cases

This approach has the following benefits:

1. Source files can run without compilation
2. JSDoc types do not provide intellisense nor type checking for typescript consumers
3. TypeDoc for documentation seems nicer than JSDoc
4. Going all in on TypeScript may be too much to ask for, for many module authors - as typescript projects require types for all dependencies, and accurate mappings, which some of the time, is just too inconvenient to bother with - https://stackoverflow.com/q/53612877/130638

Alternatives to this would be:

1. Have typescript projects support intellisense and type checking for jsdoc js files
2. Automatically generate .d.ts files for jsdoc js files

For now, the best workaround seems to be:

1. Spend weeks upon weeks converting your hundreds of modules to typescript to satisfy typescript's need for types for everything
2. Provide jsdoc .js files, and manual .d.ts files

## Examples

- https://github.com/bevry/typechecker/tree/dev-typescript-dts
- https://github.com/bevry/eachr/tree/dev-typescript-dts

## Checklist

My suggestion meets these guidelines:

* [ ] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [ ] This wouldn't change the runtime behavior of existing JavaScript code
* [ ] This could be implemented without emitting different JS based on the types of the expressions
* [ ] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, etc.)
* [ ] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

**Comment by DanielRosenwasser** [comment]
> JSDoc types do not provide intellisense nor type checking for typescript consumers

You mean when you publish to npm?

**Comment by balupton** [comment]
> You mean when you publish to npm?

Yes.

Take the package https://github.com/bevry/badges

If I consume it via a `.js` file, I get intellisense with the typing and docs provided by its jsdoc comments:

<img width="774" alt="screen shot 2018-12-18 at 11 05 37 am" src="https://user-images.githubusercontent.com/61148/50123504-1bc8ee80-02b5-11e9-92a6-86b9f2fbb2c7.png">

However, if I do it via typescript, I get nothing:

<img width="321" alt="screen shot 2018-12-18 at 11 05 55 am" src="https://user-images.githubusercontent.com/61148/50123514-284d4700-02b5-11e9-9283-91f7f960abcd.png">

Only an error whining that it needs a declaration file:

```
Could not find a declaration file for module 'badges'. '/Users/balupton/Projects/active/typescript-consume-jsdoc/node_modules/badges/index.js' implicitly has an 'any' type.
	Try`npm install @types/badges` if it exists or add a new declaration(.d.ts) file containing`declare module 'badges';`[7016]
```

Which is infuriating, especially because `@ts-check` with .js files checks types fine:

<img width="681" alt="screen shot 2018-12-18 at 11 08 52 am" src="https://user-images.githubusercontent.com/61148/50123605-80844900-02b5-11e9-8191-2e951f58a67d.png">

Which makes it so that, .ts files are actually more crippled than their .js alternatives.

**Comment by balupton** [comment]
> Sorta - try `maxNodeModuleJsDepth 1` with `allowJs`.

Interesting.

Setting it to 1 doesn't provide anything. `2` works, but doesn't provide details for `require('badges').badges`, and `3` provides details for everything.

How come this value isn't say `5` or `10` by default? As this functionality seems incredibly useful for developers.

**Comment by DanielRosenwasser** [comment]
> How come this value isn't say `5` or `10` by default? As this functionality seems incredibly useful for developers.

Type-checking everything deeply in `node_modules` is expensive, especially when `.js` files have a lot more code than `.d.ts` files.

**Comment by balupton** [comment]
> Type-checking everything deeply in `node_modules` is expensive, especially when `.js` files have a lot more code than `.d.ts` files.

Hrmm. Personally, I would prefer the increased functionality and usability, over the performance, and if performance was getting me down, then I could tune it.

As the lack of jsdoc typing for typescript projects is something that has surprised myself and several coworkers and friends.

Perhaps if documentation of `maxNodeModuleJsDepth` was mentioned more through the docs, so there is increased awareness of this behaviour, especially on the https://www.typescriptlang.org/docs/handbook/type-checking-javascript-files.html page, and in the descriptions of the commonly known `allowJs` property, then the increased awareness would workaround this expressed preference for functionality over speed.

---

That said, `maxNodeModuleJsDepth` is a suitable workaround for me. However, I imagine I need to make it known to the consumers of my packages via our README files. So they can adapt their tsconfigs appropriately, which is a bit of burden.

---

So to summarise the options for the use case for JavaScript package authors interoperating with typescript consumers:

1. Use `.js` with jsdoc, inform users they should customise their `maxNodeModuleJsDepth` property in their `tsconfig.json` files
2. Enable `.js` and `.ts` files to gain their type information from their corresponding `.d.ts` file (this issues feature request)
3. Auto-generate `.d.ts` files from the manually written jsdoc: #7546 
4. Manually write `.d.ts` files alongside the manually written jsdoc
5. Convert everything to `.ts` files

**Comment by balupton** [comment]
For what it is worth, I've updated https://github.com/bevry/projectz to automatically generate an installation guidance of maxNodeModuleJsDepth to 5 if the project it is running against has jsdoc in its dependencies.

The result of this generation can be seem on the caterpillar package for example:
https://github.com/bevry/caterpillar

It is unfortunate that such guidance is necessary.

**Comment by lo-corp** [comment]
This is a must have feature. Flow has it.

**Comment by ArsenArsen** [comment]
having an option to, for a given `foo.js` and `foo.d.ts`, treat each definition as if it was declared with the corresponding type from `foo.d.ts` would be nice.  I'm currently working around this by doing:

```ts
/** @filename foo.d.ts */
declare const foo: { a: number, b: string };

/** @filename foo.js */
/** @import * as types from "./foo" */

/** @type {typeof types.foo} */
const foo = {}; /* emits an error.  */
```

... and it feels like that's something the compiler can do

EDIT: of course, this also doesn't complain about missing definitions.  
EDIT2: further, it'd be useful to have a sort of "declaration merging", where the deductions from the `js` are used to complement the explicit specifications in the `.d.ts`, to minimize repetition

**Comment by ustun** [comment]
Thanks for bumping this @ArsenArsen but I'm unsubscribing, since this thread showed me how old I have become :D 

Damn, time passes so quickly. I remember making this comment in the sibling issue like yesterday https://github.com/microsoft/TypeScript/issues/30304#issuecomment-503971751
